R/credentials.R

```diff
@@ -0,0 +1,365 @@
+  if (!(use_ssl %in% c(TRUE, FALSE))) {
+    stop("The value supplied to `use_ssl` must be TRUE or FALSE.")
+  }
+
+  if (!is.null(provider) &&
+      provider %in% (smtp_settings() %>% dplyr::pull(short_name))) {
+
+    # Extract the record for the SMTP provider
+    settings_record <-
+      smtp_settings() %>%
+      dplyr::filter(short_name == provider)
+
+    # TODO: only set these if the incoming values are not NULL
+    # Extract settings for the provider
+    host <- settings_record$server
+    port <- settings_record$port
+    use_ssl <- settings_record$use_ssl
+  }
+
+  # If the `sender_name` isn't provided, use
+  # an empty string
+  if (is.null(sender_name)) sender_name <- ""
+
+  # Use the provider name if it's given as the
+  # `key_name` (if that's not given)
+  if (is.null(key_name) && !is.null(provider)) {
+    key_name <- provider
+  } else if (is.null(key_name)) {
+    key_name <- ""
+  }
+
+  # Create the `service_name` for `keyring`
+  service_name <- paste0("blastula-v", keyring_schema_version, "-", key_name)
+
+  credentials_list <-
+    list(
+      version = keyring_schema_version,
+      sender_name = sender_name,
+      host = host,
+      port = port,
+      use_ssl = use_ssl,
+      user = user,
+      password = password %||% getPass::getPass("Enter the SMTP server password: ")
+    )
+
+  serialized <-
+    credentials_list %>%
+    jsonlite::serializeJSON() %>%
+    as.character()
+
+  # Set the key in the system's default keyring
+  keyring::key_set_with_value(
+    service = service_name,
+    user = user,
+    password = serialized
+  )
+
+  # Issue a message stating that the file has been created
+  message("The system key store has been updated with the (`", service_name, "`) key.")
+}
+
+#' Utility function for obtaining keyring entries related to blastula creds
+#'
+#' @import keyring
+#' @importFrom dplyr as_tibble filter mutate
+#' @importFrom tidyr separate
+#' @noRd
+get_keyring_creds_table <- function(version = keyring_schema_version) {
+
+  creds_tbl <-
+    keyring::key_list() %>%
+    dplyr::as_tibble() %>%
+    dplyr::filter(grepl(paste0("blastula-v", version), service))
+
+  if (nrow(creds_tbl) == 0) {
+
+    empty_creds_tbl <-
+      dplyr::tibble(
+        key_name = NA_character_,
+        version = NA_integer_,
+        username = NA_character_
+      )[-1, ]
+
+    return(empty_creds_tbl)
+
+  } else {
+
+    creds_tbl <-
+      creds_tbl %>%
+      tidyr::separate(
+        col = service,
+        into = c("package", "version", "key_name"),
+        sep = "-",
+        remove = FALSE
+      ) %>%
+      dplyr::mutate(
+        version = version %>%
+          tidy_gsub("v", "") %>%
+          as.integer()) %>%
+      dplyr::select(service, key_name, version, username)
+  }
+
+  creds_tbl
+}
+
+#' Retrieve metadata and authentication values from keyring data
+#'
+#' @noRd
+get_smtp_keyring_creds <- function(key_name = NULL) {
+
+  creds_tbl <- get_keyring_creds_table()
```

Comment:
OK, here's where I want to ask, why not just fetch the specific key you want instead of the whole table?

Reply:
I made some changes to make this more concise (and a few explanatory comments were also added). Here we are fetching all records that have been generated by the `create_smtp_creds_key()` function then filtering by the supplied `key_name` to hopefully only get a single matching record.

If there are no records, we can `stop()` the function (with a useful message). Otherwise, we obtain the value and unserialize.